server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -116,6 +116,29 @@ public T next() {
         }
     }
 
+    public static <T, U> Iterator<U> map(Iterator<? extends T> input, Function<T, U> fn) {
+        if (input.hasNext()) {
+            return new MapIterator<>(input, fn);
+        } else {
+            return Collections.emptyIterator();
+        }
+    }
+
+    private record MapIterator<T, U>(Iterator<? extends T> input, Function<T, U> fn) implements Iterator<U> {
```

Comment:
What about `remove` and `forEachRemaining`?

Reply:
Could delegate `forEachRemaining` with `input.forEachRemaining(f.compose(fn))`

Reply:
The default implementations are fine. We don't use them anywhere at the moment anyway, so this'd just be dead code.

Reply:
This is a general-purpose implementation, so it at least should have a delegating implementation of `remove` - the default just throws

Reply:
I disagree - we can easily add such a thing if ever we need it, but we don't need it today (and I do not expect we'll need it in future either). You could make the same argument for the other utility implementations here.